### PR TITLE
[TTI][RISCV] Model cost of loading constants arms of selects and compares

### PR DESCRIPTION
**Pull request by preames**
This follows in the spirit of 7d82c99403f615f6236334e698720bf979959704, and extends the costing API for compares and selects to provide information about the operands passed in an analogous manner.  This allows us to model the cost of materializing the vector constant, as some select-of-constants are significantly more expensive than others when you account for the cost of materializing the constants involved.

This is a stepping stone towards fixing https://github.com/llvm/llvm-project/issues/109466.  A separate SLP patch will be required to utilize the new API.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 1]
```diff
@@ -1618,23 +1618,38 @@ InstructionCost RISCVTTIImpl::getMemoryOpCost(unsigned Opcode, Type *Src,
...
-                                     I);
+                                     Op1Info, Op2Info, I);
 
   // Skip if scalar size of ValTy is bigger than ELEN.
   if (ValTy->isVectorTy() && ValTy->getScalarSizeInBits() > ST->getELen())
     return BaseT::getCmpSelInstrCost(Opcode, ValTy, CondTy, VecPred, CostKind,
-                                     I);
+                                     Op1Info, Op2Info, I);
+
+  auto getConstantMatCost =
```
Comment (alexey-bataev) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1638]: GetConstantMatCost

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1638]: From the coding standards:

"Variable names should be nouns (as they represent state). The name should be camel case, and start with an upper case letter (e.g. Leader or Boats).

Function names should be verb phrases (as they represent actions), and command-like function should be imperative. The name should be camel case, and start with a lower case letter (e.g. openFile() or isFoo())"

To me, a lambda is a function, not a variable.  I remember this being discussed extensively at some point in the past, but I could not find clear guidance in the coding standards.  I'd prefer to leave this as is since it matches the naming elsewhere in this file.

Comment (alexey-bataev) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1638]: clang-tidy suggests using capital letters for lambdas, at least. Technically, they are variables

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -10760,11 +10764,14 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
             TTIRef.getArithmeticInstrCost(E->getAltOpcode(), VecTy, CostKind);
       } else if (auto *CI0 = dyn_cast<CmpInst>(VL0)) {
         auto *MaskTy = getWidenedType(Builder.getInt1Ty(), VL.size());
-        VecCost = TTIRef.getCmpSelInstrCost(E->getOpcode(), VecTy, MaskTy,
-                                            CI0->getPredicate(), CostKind, VL0);
+        VecCost = TTIRef.getCmpSelInstrCost(
+            E->getOpcode(), VecTy, MaskTy, CI0->getPredicate(), CostKind,
+            {TTI::OK_AnyValue, TTI::OP_None}, {TTI::OK_AnyValue, TTI::OP_None},
```
Comment (lukel97) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:10769]: Do we need to use getOperandInfo on VL0 here?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 1 above)
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Found the prior discussion: https://discourse.llvm.org/t/naming-lambdas-in-llvm-coding-standards/62689/9

Seems like the general consensus was CamelCase, so I changed.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 2 above)
Comment (preames) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:10769]: Need to?  No.  It's not correctness related.  However doing so would be likely reasonable.  I'd plan to do a follow up to adjust other call sites, I started with the subset required by the original test and printing the cost model.

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:10769]: Yes, better to start with the generic approach and tune then in the follow up patch(es)

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -10519,8 +10521,10 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
     auto GetVectorCost = [&](InstructionCost CommonCost) {
       auto *MaskTy = getWidenedType(Builder.getInt1Ty(), VL.size());
 
-      InstructionCost VecCost = TTI->getCmpSelInstrCost(
-          E->getOpcode(), VecTy, MaskTy, VecPred, CostKind, VL0);
+      InstructionCost VecCost =
+          TTI->getCmpSelInstrCost(E->getOpcode(), VecTy, MaskTy, VecPred,
+                                  CostKind, getOperandInfo(E->getOperand(0)),
+                                  getOperandInfo(E->getOperand(1)), VL0);
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:10527]: Maybe use generic data here too to add full support later along with the tests?

Comment (preames) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:10527]: Can you rephrase?  I can't figure out your intent.  My best guess is that you want me to split the patch into an SLP portion and a TTI portion?  Is that what you meant?

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:10527]: Kind of. Use default values here and use actual operand info in a separate patch

Comment (preames) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: Done.

**Review by alexey-bataev — APPROVED** [review]
LG
